components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/mediasession/GeckoMediaSessionDelegate.kt

```diff
@@ -30,11 +44,22 @@ internal class GeckoMediaSessionDelegate(
         mediaSession: GeckoViewMediaSession,
         metaData: GeckoViewMediaSession.Metadata
     ) {
-        engineSession.notifyObservers {
-            onMediaMetadataChanged(
-                MediaSession.Metadata(metaData.title, metaData.artist, metaData.album)
-            )
+        CoroutineScope(Dispatchers.IO).launch(artWorkDispatcher) {
+            val artwork = metaData.artwork?.getBitmap(ARTWORK_IMAGE_SIZE)?.await()
+
+            engineSession.notifyObservers {
+                onMediaMetadataChanged(
+                    MediaSession.Metadata(metaData.title, metaData.artist, metaData.album, artwork)
+                )
+            }
+        }
```

Comment:
I think there's a problem here:

* Every `GeckoMediaSessionDelegate` will spin up its own thread ... and it looks like those are never cleaned up?
* onMediaMetadataChanged() is now delayed until we were finished fetching the bitmap?
* The delay may introduce races where we invoke `onMediaMetadataChanged()` after the media has already changed?
* We may be prematurely fetching all artwork and keeping it in memory. (Maybe less a concern since we need them almost always, I guess?)

Maybe we can design it like "browser actions" in WebExtensions. The action class exposes a lambda (`val loadIcon: (suspend (Int) -> Bitmap?)?`). With that we can construct and forward the data immediately. And only code that needs the `Bitmap` needs to figure out how/when to fetch by invoking the lambda. What do you think?

Reply:
Sure, I'll take a look at `loadIcon` and see if I can use the same pattern.  Thanks,

Reply:
Updated to address your concerns.  onMediaMetadataChanged() is not blocked anymore.  Please let me know if this is what you have in mind.  Thanks,